src/experiments/tcpsocket/api.js

```diff
@@ -0,0 +1,91 @@
+"use strict";
+/* exported tcpsocket */
+/* global ExtensionAPI, ChromeUtils, Cu */
+
+const { TCPSocket } = Cu.getGlobalForObject(
+    ChromeUtils.import("resource://gre/modules/Services.jsm")
+);
+
+/**
+ * Concatenate two Uint8Array objects
+ */
+function concatUint8Arrays(a, b) {
+    let newArr = new Uint8Array(a.length + b.length);
+    newArr.set(a, 0);
+    newArr.set(b, a.length);
+    return newArr;
+}
+
+var tcpsocket = class tcpsocket extends ExtensionAPI {
+    getAPI(context) {
+        return {
+            experiments: {
+                tcpsocket: {
+                    /** 
+                     * Send a DNS query stored in buf over a TCP socket to a 
+                     * nameserver addressed by addr
+                     */
+                    async sendDNSQuery(addr, buf) {
+                        let tcp_socket;
+
+                        try {
+                            // Wait until the socket is open before sending data.
+                            // If we get an 'error' event before an 'open' event, 
+                            // throw an ExtensionError.
+                            tcp_socket = new TCPSocket(addr, 53, { binaryType: "arraybuffer" });
+                            let eventType = await new Promise((resolve, reject) => {
+                                tcp_socket.onopen = ((event) => {
+                                    resolve(event.type);
+                                });
+
+                                tcp_socket.onerror = ((event) => {
+                                    throw new ExtensionError(event.name);
+                                });
+                            });
+                            if (eventType != "open" || tcp_socket.readyState != "open") {
+                                throw new ExtensionError("TCP socket didn't properly open");
+                            }
+
+                            // After we know that the socket is open, send the bytes
+                            tcp_socket.send(buf.buffer, buf.byteOffset, buf.byteLength);
+                            let responseBytes = await new Promise((resolve, reject) => {
+                                let data = new Uint8Array();
+                                let expectedLength;
+
+                                tcp_socket.ondata = ((event) => {
+                                    if (data.length == 0) {
+                                        expectedLength = new DataView(event.data).getUint16(0);
+                                        let receivedData = new Uint8Array(event.data);
+                                        data = receivedData;
+                                    } else {
+                                        data = concatUint8Arrays(data, receivedData);
```

Comment:
Since `data` is already initialized to be a `Uint8Array`, you can unconditionally call the following:

```
data = concatUint8Arrays(data, new Uint8Array(event.data));
```

followed by

```
if (data.length >= 2 && !expectedLength) {
  expectedLength = new DataView(data).getUint16(0) + 2;
  // TODO: Maybe a bound check to avoid excessive values?
}
if (data.length >= expectLength) {
  // Maybe reject() if you want to ignore responses that are too large.
  resolve(data);
}
```

Reply:
Addressed by 7e91fde